Userland/Libraries/LibGPU/Driver.cpp

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright (c) 2022, Stephan Unverwerth <s.unverwerth@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/HashMap.h>
+#include <AK/String.h>
+#include <AK/WeakPtr.h>
+#ifdef __serenity__
+#    include <LibDl/dlfcn.h>
+#else
+#    include <dlfcn.h>
+#endif
+#include <LibGPU/Driver.h>
+
+namespace GPU {
+
+// FIXME: Think of a better way to configure these paths. Maybe use ConfigServer?
+static HashMap<String, String> const s_driver_path_map
+{
+#if defined(__serenity__)
+    { "softgpu", "/usr/lib/libsoftgpu.so" },
+#elif defined(__APPLE__)
+    { "softgpu", "./liblagom-softgpu.dylib" },
+#else
+    { "softgpu", "./liblagom-softgpu.so" },
+#endif
+};
+
+static HashMap<String, WeakPtr<Driver>> s_loaded_drivers;
+
+ErrorOr<NonnullRefPtr<Driver>> Driver::try_create(StringView driver_name)
+{
+    // Check if the library for this driver is already loaded
+    auto already_loaded_driver = s_loaded_drivers.find(driver_name);
+    if (already_loaded_driver != s_loaded_drivers.end() && !already_loaded_driver->value.is_null())
+        return *already_loaded_driver->value;
+
+    // Nope, we need to load the library
+    auto it = s_driver_path_map.find(driver_name);
+    if (it == s_driver_path_map.end())
+        return Error::from_string_literal("The requested GPU driver was not found in the list of allowed driver libraries");
+
+    auto lib = dlopen(it->value.characters(), RTLD_NOW);
+    if (!lib)
+        return Error::from_string_literal("The library for the requested GPU driver could not be opened");
+
+    auto serenity_gpu_create_device = reinterpret_cast<serenity_gpu_create_device_t>(dlsym(lib, "serenity_gpu_create_device"));
+    if (!serenity_gpu_create_device) {
+        dlclose(lib);
```

Comment:
Can this just be wrapped in an RAII class?

Reply:
It is already. If you look a few lines below we inject this into the `Driver` class, but this class can only be instantiated if we have a library that follows our spec.